File: server/servers/api/src/main/scala/com/prisma/api/schema/FilterArguments.scala
package com.prisma.api.schema

import com.prisma.shared.models.{Field, Model, TypeIdentifier}

case class FieldFilterTuple(field: Option[Field], filterArg: FilterArgument)
case class FilterArgument(name: String, description: String, isList: Boolean = false)

class FilterArguments(model: Model, isSubscriptionFilter: Boolean = false) {

  private val index = model.fields
    .flatMap(field => {
      FilterArguments
        .getFieldFilters(field)
        .map(filter => { (field.name + filter.name, FieldFilterTuple(Some(field), filter)) })
    })
    .toMap

  def lookup(filter: String): FieldFilterTuple = filter match {
    case "AND" =>
      FieldFilterTuple(None, FilterArguments.ANDFilter)

    case "OR" =>
      FieldFilterTuple(None, FilterArguments.ORFilter)

    case "NOT" =>
      FieldFilterTuple(None, FilterArguments.NOTFilter)

    case "boolean" if isSubscriptionFilter =>
      FieldFilterTuple(None, FilterArguments.booleanFilter)

    case "node" if isSubscriptionFilter =>
      FieldFilterTuple(None, FilterArguments.nodeFilter)

    case _ =>
      index.get(filter) match {
        case None                   => throw new Exception(s""""No field for the filter "$filter" has been found.""")
        case Some(fieldFilterTuple) => fieldFilterTuple
      }
  }
}

object FilterArguments {

  val ANDFilter     = FilterArgument("AND", "Logical AND on all given filters.")
  val ORFilter      = FilterArgument("OR", "Logical OR on all given filters.")
  val NOTFilter     = FilterArgument("NOT", "Logical NOT on all given filters combined by AND.")
  val booleanFilter = FilterArgument("boolean", "")
  val nodeFilter    = FilterArgument("node", "")

  private val baseFilters = List(
    FilterArgument("", ""),
    FilterArgument("_not", "All values that are not equal to given value.")
  )

  private val inclusionFilters = List(
    FilterArgument("_in", "All values that are contained in given list.", isList = true),
    FilterArgument("_not_in", "All values that are not contained in given list.", isList = true)
  )

  private val alphanumericFilters = List(
    FilterArgument("_lt", "All values less than the given value."),
    FilterArgument("_lte", "All values less than or equal the given value."),
    FilterArgument("_gt", "All values greater than the given value."),
    FilterArgument("_gte", "All values greater than or equal the given value.")
  )

  private val stringFilters = List(
    FilterArgument("_contains", "All values containing the given string."),
    FilterArgument("_not_contains", "All values not containing the given string."),
    FilterArgument("_starts_with", "All values starting with the given string."),
    FilterArgument("_not_starts_with", "All values not starting with the given string."),
    FilterArgument("_ends_with", "All values ending with the given string."),
    FilterArgument("_not_ends_with", "All values not ending with the given string.")
  )

  private val multiRelationFilters = List(
    FilterArgument("_every", "All nodes where all nodes in the relation satisfy the given condition."),
    FilterArgument("_some", "All nodes that have at least one node in the relation satisfying the given condition."),
    FilterArgument("_none", "All nodes that have no node in the relation satisfying the given condition.")
  )

  private val oneRelationFilters = List(FilterArgument("", ""))

  def getFieldFilters(field: Field): List[FilterArgument] = {
    val filters =
      if (field.isList) {
        field.typeIdentifier match {
          case TypeIdentifier.Relation => List(multiRelationFilters)
          case _                       => List()
        }
      } else {
        field.typeIdentifier match {
          case TypeIdentifier.UUID     => List(baseFilters, inclusionFilters)
          case TypeIdentifier.Cuid     => List(baseFilters, inclusionFilters, alphanumericFilters, stringFilters)
          case TypeIdentifier.String   => List(baseFilters, inclusionFilters, alphanumericFilters, stringFilters)
          case TypeIdentifier.Int      => List(baseFilters, inclusionFilters, alphanumericFilters)
          case TypeIdentifier.Float    => List(baseFilters, inclusionFilters, alphanumericFilters)
          case TypeIdentifier.Boolean  => List(baseFilters)
          case TypeIdentifier.Enum     => List(baseFilters, inclusionFilters)
          case TypeIdentifier.DateTime => List(baseFilters, inclusionFilters, alphanumericFilters)
          case TypeIdentifier.Json     => List()
          case TypeIdentifier.Relation => List(oneRelationFilters)
          case _                       => List()
        }
      }

    filters.flatten
  }
}

File: server/prisma-rs/query-engine/core/src/schema/builders/filter_arguments.rs
use prisma_models::{Field, TypeIdentifier};

/// (Filter suffix, is list)
pub struct FilterArgument {
    pub suffix: &'static str,
    pub is_list: bool,
}

/// Wrapper type for static initialization (circumventing lazy_static type issues).
struct StaticFilterArguments {
    pub base: Vec<FilterArgument>,
    pub inclusion: Vec<FilterArgument>,
    pub alphanumeric: Vec<FilterArgument>,
    pub string: Vec<FilterArgument>,
    pub multi_relation: Vec<FilterArgument>,
    pub one_relation: Vec<FilterArgument>,
}

lazy_static! {
    static ref filter_arguments: StaticFilterArguments = StaticFilterArguments {
        base: vec![
            FilterArgument { suffix: "", is_list: false },
            FilterArgument { suffix: "_not",is_list: false }, // All values that are not equal to given value.
        ],

        inclusion: vec![
            FilterArgument { suffix: "_in",is_list: true }, // All values that are contained in given list.
            FilterArgument { suffix: "_not_in",is_list: true } // All values that are not contained in given list.
        ],

        alphanumeric: vec![
            FilterArgument { suffix: "_lt",is_list: false }, // All values less than the given value.
            FilterArgument { suffix: "_lte",is_list: false }, // All values less than or equal the given value.
            FilterArgument { suffix: "_gt",is_list: false }, // All values greater than the given value.
            FilterArgument { suffix: "_gte",is_list: false } // All values greater than or equal the given value.
        ],

        string: vec![
            FilterArgument { suffix: "_contains",is_list: false }, // All values containing the given string.
            FilterArgument { suffix: "_not_contains",is_list: false }, // All values not containing the given string.
            FilterArgument { suffix: "_starts_with",is_list: false }, // All values starting with the given string.
            FilterArgument { suffix: "_not_starts_with",is_list: false }, // All values not starting with the given string.
            FilterArgument { suffix: "_ends_with",is_list: false }, // All values ending with the given string.
            FilterArgument { suffix: "_not_ends_with",is_list: false } // All values not ending with the given string.
        ],

        multi_relation: vec![
            FilterArgument { suffix: "_every",is_list: false }, // All nodes where all nodes in the relation satisfy the given condition.
            FilterArgument { suffix: "_some",is_list: false }, // All nodes that have at least one node in the relation satisfying the given condition.
            FilterArgument { suffix: "_none",is_list: false } // All nodes that have no node in the relation satisfying the given condition.
        ],

        one_relation: vec![FilterArgument { suffix: "", is_list: false }],
    };
}

pub fn get_field_filters<'a>(field: &Field) -> Vec<&'a FilterArgument> {
    let args = &filter_arguments;

    if field.is_list() {
        match field.type_identifier() {
            TypeIdentifier::Relation => args.multi_relation.iter().collect(),
            _ => vec![],
        }
    } else {
        let filters = match field.type_identifier() {
            TypeIdentifier::UUID => vec![],
            TypeIdentifier::GraphQLID => vec![&args.base, &args.inclusion, &args.alphanumeric, &args.string],
            TypeIdentifier::String => vec![&args.base, &args.inclusion, &args.alphanumeric, &args.string],
            TypeIdentifier::Int => vec![&args.base, &args.inclusion, &args.alphanumeric],
            TypeIdentifier::Float => vec![&args.base, &args.inclusion, &args.alphanumeric],
            TypeIdentifier::Boolean => vec![&args.base],
            TypeIdentifier::Enum => vec![&args.base, &args.inclusion],
            TypeIdentifier::DateTime => vec![&args.base, &args.inclusion, &args.alphanumeric],
            TypeIdentifier::Json => vec![],
            TypeIdentifier::Relation => vec![&args.one_relation],
        };

        filters
            .into_iter()
            .map(|l| l.iter().collect::<Vec<&'a FilterArgument>>())
            .flatten()
            .collect()
    }
}
